plugins/crypto/mbedtls/securitypolicy_mbedtls_common.c

```diff
@@ -289,6 +300,252 @@ mbedtls_decrypt_rsaOaep(mbedtls_pk_context *localPrivateKey,
+                                                     MBEDTLS_X509_KU_NON_REPUDIATION |
+                                                     MBEDTLS_X509_KU_KEY_ENCIPHERMENT) != 0) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    /* Add entropy */
+    if(entropyContext && nonce && nonce->length > 0) {
+        if(mbedtls_entropy_update_manual(entropyContext,
+                                         (const unsigned char*)(nonce->data),
+                                         nonce->length) != 0) {
+            retval = UA_STATUSCODE_BADINTERNALERROR;
+            goto cleanup;
+        }
+    }
+
+    /* Get subject from argument or read it from certificate */
+    if(subjectName && subjectName->length > 0) {
+        /* subject from argument */
+        subj = (char *)UA_malloc(subjectName->length + 1);
+        if(!subj) {
+            retval = UA_STATUSCODE_BADOUTOFMEMORY;
+            goto cleanup;
+        }
+        memset(subj, 0x00, subjectName->length + 1);
+        strncpy(subj, (char *)subjectName->data, subjectName->length);
+        /* search for / in subject and replace it by comma */
+        char *p = subj;
+        for(size_t i = 0; i < subjectName->length; i++) {
+            if(*p == '/' ) {
+                *p = ',';
+            }
+            ++p;
+        }
+    } else {
+        /* read subject from certificate */
+        mbedtls_x509_name s = x509Cert.subject;
+        subj = (char *)UA_malloc(UA_MAXSUBJECTLENGTH);
+        if(!subj) {
+            retval = UA_STATUSCODE_BADOUTOFMEMORY;
+            goto cleanup;
+        }
+        if(mbedtls_x509_dn_gets(subj, UA_MAXSUBJECTLENGTH, &s) <= 0) {
+            retval = UA_STATUSCODE_BADINTERNALERROR;
+            goto cleanup;
+        }
+    }
+
+    /* Set the subject in CSR context */
+    if(mbedtls_x509write_csr_set_subject_name(&request, subj) != 0) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    /* Get the subject alternate names from certificate and set them in CSR context*/
+    const mbedtls_x509_sequence *san_list = &x509Cert.subject_alt_names;
+    mbedtls_x509write_csrSetSubjectAltName(&request, san_list);
+
+    /* Set private key in CSR context */
+    if(newPrivateKey) {
+        mbedtls_pk_init(csrLocalPrivateKey);
+        mbedtls_pk_setup(csrLocalPrivateKey, mbedtls_pk_info_from_type(MBEDTLS_PK_RSA));
+
+        size_t keySize = 0;
+        UA_CertificateUtils_getKeySize(&securityPolicy->localCertificate, &keySize);
+        mbedtls_rsa_gen_key(mbedtls_pk_rsa(*csrLocalPrivateKey), mbedtls_ctr_drbg_random,
+                            drbgContext, (unsigned int)keySize, 65537);
+        mbedtls_x509write_csr_set_key(&request, csrLocalPrivateKey);
+        mbedtls_writePrivateKeyDer(csrLocalPrivateKey, newPrivateKey);
+    } else {
+        mbedtls_x509write_csr_set_key(&request, localPrivateKey);
+    }
+
+
+    unsigned char requestBuf[4096];
```

Comment:
Magic number 4096.
Can this become a problem?
You can make this a #define at the top of the file.